cub/test/thread_reduce/catch2_test_thread_reduce_partial.cu

```diff
@@ -0,0 +1,611 @@
+  const int num_items          = GENERATE_COPY(take(3, random(1, max_size)));
+  const int valid_items        = GENERATE_COPY(
+    take(1, random(2, ::cuda::std::max(2, num_items - 1))),
+    take(1, random(num_items + 2, ::cuda::std::numeric_limits<int>::max())),
+    values({1, num_items, num_items + 1}));
+  c2h::device_vector<value_t> d_in(num_items);
+  c2h::device_vector<accum_t> d_out(1);
+  c2h::gen(C2H_SEED(num_seeds), d_in, dist_param::min(), dist_param::max());
+  c2h::host_vector<value_t> h_in = d_in;
+  CAPTURE(h_in, dist_param::min(), dist_param::max());
+  CAPTURE(c2h::type_name<value_t>(), c2h::type_name<decltype(reduce_op)>(), valid_items, num_items, operator_identity);
+  const int bounded_valid_items = ::cuda::std::min(valid_items, num_items);
+  const value_t reference_result =
+    compute_single_problem_reference(h_in.cbegin(), h_in.cbegin() + bounded_valid_items, reduce_op, operator_identity);
+
+  run_thread_reduce_partial_kernel(num_items, d_in, d_out, reduce_op, valid_items);
+  REQUIRE(reference_result == c2h::host_vector<accum_t>(d_out)[0]);
+}
+
+#endif // TEST_HALF_T() || TEST_BF_T()
+
+C2H_TEST("ThreadReduce Container Tests", "[reduce][thread]")
+{
+  using op_t       = ::cuda::std::plus<>;
+  using dist_param = dist_interval<int, op_t>;
+  c2h::device_vector<int> d_in(max_size);
+  c2h::device_vector<int> d_out(1);
+  c2h::gen(C2H_SEED(num_seeds), d_in, dist_param::min(), dist_param::max());
+  c2h::host_vector<int> h_in = d_in;
+  const int valid_items      = GENERATE_COPY(
+    take(1, random(2, max_size - 2)),
+    take(1, random(max_size + 2, ::cuda::std::numeric_limits<int>::max())),
+    values({1, max_size - 1, max_size, max_size + 1}));
+  const int bounded_valid_items = ::cuda::std::clamp(valid_items, 0, max_size);
+  auto reference_result =
+    compute_single_problem_reference(h_in.cbegin(), h_in.cbegin() + bounded_valid_items, op_t{}, 0);
+
+  thread_reduce_partial_kernel_array<max_size>
+    <<<1, 1>>>(thrust::raw_pointer_cast(d_in.data()), thrust::raw_pointer_cast(d_out.data()), op_t{}, valid_items);
+  REQUIRE(cudaSuccess == cudaPeekAtLastError());
+  REQUIRE(cudaSuccess == cudaDeviceSynchronize());
+  REQUIRE(reference_result == c2h::host_vector<int>(d_out)[0]);
+
+  thread_reduce_partial_kernel_span<max_size>
+    <<<1, 1>>>(thrust::raw_pointer_cast(d_in.data()), thrust::raw_pointer_cast(d_out.data()), op_t{}, valid_items);
+  REQUIRE(cudaSuccess == cudaPeekAtLastError());
+  REQUIRE(cudaSuccess == cudaDeviceSynchronize());
+  REQUIRE(reference_result == c2h::host_vector<int>(d_out)[0]);
+
+#if _CCCL_STD_VER >= 2023
+  thread_reduce_partial_kernel_mdspan<max_size>
+    <<<1, 1>>>(thrust::raw_pointer_cast(d_in.data()), thrust::raw_pointer_cast(d_out.data()), op_t{}, valid_items);
+  REQUIRE(cudaSuccess == cudaPeekAtLastError());
+  REQUIRE(cudaSuccess == cudaDeviceSynchronize());
+  REQUIRE(reference_result == c2h::host_vector<int>(d_out)[0]);
+#endif // _CCCL_STD_VER >= 2023
+}
+
+struct segment
```

Comment:
The following `segment` class and operators on it are also duplicated. They differ from the more extensive version (multiple counters vs one flag) in the `DeviceScan` test. But once we have a `DeviceScan` that guarantees not to apply the binary op to invalid items, I think we could use this more lean version even there because regressions should be easy enough to trace back to a specific root cause given a specific PR.

Reply:
Already done in #5300. Waiting for it to get merged?